# Домашнее задание на проверку знаний библиотеки для анализа данных Pandas.

Вам представлен датасет, который содержит информацию по результатам сезона 2016 года игры в бейсбол.

Вам необходимо загрузить датасет (он находится в архиве с текущим домашним заданием). Провести его очистку, если это необходимо и ответить на поставленные аналитические
вопросы.

Список колонок, представленных в датасете:

---
* attendance        - количество зрителей матча
* away_team         - название гостевой команды
* away_team_errors  - количество ошибок гостевой команды
* away_team_hits    - количество хитов гостевой команды (хит в бейсболе - удар при котором отбивающий достиг первой базы)
* away_team_runs    - количество ранов гостевой команды (ран в бейсболе - очко заработанное игроком нападения)
* date              - дата проведения матча
* field_type        - тип игрового поля
* game_type         - тип игры
* home_team         - название домашней команды
* home_team_errors  - количество ошибок домашней команды
* home_team_hits    - количество хитов домашней команды (хит в бейсболе - удар при котором отбивающий достиг первой базы)
* home_team_runs    - количество ранов домашней команды (ран в бейсболе - очко заработанное игроком нападения)
* start_time        - время начала игрового матча
* venue             - название места проведения матча (стадион, поле, арена)
* day_of_week       - день недели проведения матча
* temperature       - температура воздуха в день проведения матча в фаренгейтах
* wind_speed        - скорость ветра в день проведения матча
* wind_direction    - направление скорости ветра
* sky               - облачность
* total_runs        - общее количество ранов для двух команд
* game_hours_dec    - длительность матча указаная в часах
* season            - тип игрового сезона
* home_team_win     - результат домашней команды (1 - выигрыш)
* home_team_loss    - результат домашней команды (0 - проигрыш)
* home_team_outcome - исход матча
---

Всего в задании 20 вопросов. За каждый правильный вопрос начисляется 5 баллов. Таким образом за все правильные ответы вы получите - 100 баллов. Оценка затем будет приведена к 10-ти бальной шкале.

---

**Баллы могут быть снижены** за качество кода - старайтесь чтобы код был максимально лаконичным. Названия переменных (если они используются) должны отражать максимальную ее сущность. Вывод старайтесь делать минимальным, так как первичная цель - сделать качественный анализ, чтобы быстро ответить на поставленный вопрос (никаких таблиц на 100 строк).

**Требования по сдаче практического задания**:
1. Ответ на вопрос и текстовый markdown с вашей логикой, почему вы выполнили это действие.
2. Залить в репу на github в папку module_1
3. Ссылку на репо отправлять мне на почту alexadevv@gmail.com не позднее дедлайна. 
4. Тема письма: Домашнее задание по модулю 1 <Фамилия Имя>
5. После дедлайна, неотправленные на почту задания, оцениваются максимум в 7 баллов.

In [3]:
# Импортируйте все необходимые библиотеки и модули которые вам необходимы
import pandas as pd
import numpy as np

In [4]:
# Загрузите датасет и проведите его очистку и форматирование если это необходимо
df = pd.read_csv(r'C:\Users\Herman\Study\Data Science Course(it-by)\homework_1\baseball_games.csv')

# 1. На какую игру пришло максимальное количество зрителей за весь сезон игр?

In [79]:
df.loc[df["attendance"].idxmax(), ["date", "home_team", "away_team", "venue", "attendance"]]


date                   2016-10-20
home_team     Los Angeles Dodgers
away_team            Chicago Cubs
venue              Dodger Stadium
attendance                54449.0
Name: 358, dtype: object

Метод [df["attendance"].idxmax()находит индекс строки, где значение "attendance" макисмальное. Полученный индекс передается в df.loc, который извлечёт соответствующую строку.

Найдя нужную строку в датасете выводим основную информацию о матче: дату, домашнюю и гостевую команды, арену, кол-во зрителей.

# 2. Какая игра была самая холодная (temperature) за весь сезон?

In [80]:
df.loc[df["temperature"].idxmin(), ["date", "home_team", "away_team", "venue", "temperature"]]

date                 2016-04-09
home_team        Detroit Tigers
away_team      New York Yankees
venue             Comerica Park
temperature                31.0
Name: 2409, dtype: object

Метод [df["temperature"].idxmin()находит индекс строки, где значение "temperature" минимальное. Полученный индекс передается в df.loc, который извлечёт соответствующую строку.

Найдя нужную строку в датасете выводим основную информацию о матче: дату, домашнюю и гостевую команды, арену, температуру.

# 3. Какая игра была самая теплая за весь сезон?

In [81]:
df.loc[df["temperature"].idxmax(), ["date", "home_team", "away_team", "venue", "temperature"]]

date                     2016-05-13
home_team      Arizona Diamondbacks
away_team      San Francisco Giants
venue                   Chase Field
temperature                   101.0
Name: 2026, dtype: object

Метод [df["temperature"].idxmax()находит индекс строки, где значение "temperature" макисмальное. Полученный индекс передается в df.loc, который извлечёт соответствующую строку.

Найдя нужную строку в датасете выводим основную информацию о матче: дату, домашнюю и гостевую команды, арену, температуру.

# 4. Какая игра в сезоне была самая долгая по продолжительности матча?

In [82]:
df.loc[df["game_hours_dec"].idxmax(),["date", "home_team", "away_team", "venue", "game_hours_dec"]]

date                     2016-07-01
home_team         Toronto Blue Jays
away_team         Cleveland Indians
venue                 Rogers Centre
game_hours_dec             6.216667
Name: 1445, dtype: object

Метод [df["attendance"].idxmax()находит индекс строки, где значение "game_hours_dec" макисмальное. Полученный индекс передается в df.loc, который извлечёт соответствующую строку.

Найдя нужную строку в датасете выводим основную информацию о матче: дату, домашнюю и гостевую команды, арену, продолжительность матча.

# 5. Какая игра в сезоне была самая короткая по продолжительности матча?

In [83]:
df.loc[df["game_hours_dec"].idxmin(),["date", "home_team", "away_team", "venue", "game_hours_dec"]]

date                      2016-09-29
home_team         Pittsburgh Pirates
away_team               Chicago Cubs
venue                       PNC Park
game_hours_dec                  1.25
Name: 423, dtype: object

Метод [df["attendance"].idxmin()находит индекс строки, где значение "game_hours_dec" минимальное. Полученный индекс передается в df.loc, который извлечёт соответствующую строку.

Найдя нужную строку в датасете выводим основную информацию о матче: дату, домашнюю и гостевую команды, арену, продолжительность матча.

# 6. Сколько матчей в сезоне закончилось ничьей?

In [84]:
int(((df["home_team_win"] == 0) & (df["home_team_loss"] == 0)).sum())

1

(df["home_team_win"] == 0) означает что домашняя команда не выиграла матч
в то же время 
df["home_team_loss"] == 0) означает что домашнее команда не проиграла матч
    
& означает, что оба условия должны выполняться одновременно

получаем True только для ничьих, после с помощью.sum() считаем кол-во ничьих и с помощью int приводим к обычному числу

# 7. Какая игра была последней в сезоне?

In [85]:
df.sort_values("date").tail(1)[["date", "home_team", "away_team", "venue"]]

date          home_team     away_team              venue
350  2016-11-02  Cleveland Indians  Chicago Cubs  Progressive Field

Для определения послднего матч используем столбец "date". Датасет сортруется по дате в хронологическом порядке, то есть от самой ранней до самой поздней и выбирается последняя строка.

Найдя нужную строку в датасете выводим основную информацию о матче: дату, домашнюю и гостевую команды, арену.

# 8. У какой игры было минимальное количество зрителей?

In [86]:
df.loc[df["attendance"].idxmin(),["date", "home_team", "away_team", "venue", "attendance"]]

date                 2016-05-04
home_team     Cleveland Indians
away_team        Detroit Tigers
venue         Progressive Field
attendance               8766.0
Name: 2130, dtype: object

Метод [df["attendance"].idxmin()находит индекс строки, где значение "attendance" минимальное. Полученный индекс передается в df.loc, который извлечёт соответствующую строку.

Найдя нужную строку в датасете выводим основную информацию о матче: дату, домашнюю и гостевую команды, арену, кол-во зрителей.

# 9. Какая игра в сезоне была самая ветренная?

In [87]:
df.loc[df["wind_speed"].idxmax(),["date", "home_team", "away_team","venue", "wind_speed"]]

date                    2016-06-13
home_team     San Francisco Giants
away_team        Milwaukee Brewers
venue                    AT&T Park
wind_speed                    25.0
Name: 1655, dtype: object

Метод [df["wind_speed"].idxmax()находит индекс строки, где значение "wind_speed" максимальное. Полученный индекс передается в df.loc, который извлечёт соответствующую строку.

Найдя нужную строку в датасете выводим основную информацию о матче: дату, домашнюю и гостевую команды, арену, скорость ветра.

# 10. В какой игре получили максимальное количество очков?

In [89]:
df.loc[df["total_runs"].idxmax(), ["date", "venue", "home_team", "away_team", "home_team_runs", "away_team_runs", "total_runs"]]

date                    2016-06-02
venue                   Petco Park
home_team         San Diego Padres
away_team         Seattle Mariners
home_team_runs                  13
away_team_runs                  16
total_runs                      29
Name: 1788, dtype: object

Метод [df["total_runs"].idxmax()находит индекс строки, где значение "total_runs" максимальное. Полученный индекс передается в df.loc, 
который извлечёт соответствующую строку.

Найдя нужную строку в датасете выводим основную информацию о матче: дату,арену, домашнюю и гостевую команды; 
очки, которые заработала каждая из команд и общее количество полученных очков.

# 11. Какая игра содержала максимальное количество ошибок домашней команды?

In [90]:
df.loc[df["home_team_errors"].idxmax(), ["date", "home_team", "away_team", "home_team_errors", "venue"]]

date                          2016-07-27
home_team              Milwaukee Brewers
away_team           Arizona Diamondbacks
home_team_errors                       5
venue                        Miller Park
Name: 1178, dtype: object

Метод [df["home_team_errors"].idxmax()находит индекс строки, где значение "home_team_errors" максимальное. Полученный индекс передается в df.loc, 
который извлечёт соответствующую строку.

Найдя нужную строку в датасете выводим основную информацию о матче: дату, домашнюю и гостевую команды, кол-во ошибок домашней команды, арену. 
очки, которые заработала каждая из команд и общее количество полученных очков.

# 12. В какой игре было максимальное количество хитов?

In [91]:
df ["total_hits"] = df["home_team_hits"] + df["away_team_hits"]
df.loc[df["total_hits"].idxmax(), ["date", "home_team", "away_team", "venue", "total_hits"]]

date              2016-07-04
home_team     Boston Red Sox
away_team      Texas Rangers
venue            Fenway Park
total_hits                37
Name: 1413, dtype: object

Для начала выяснее общее количество хитов в каждой игре, суммиируя хиты домашней и гостевой команды.

После чего метод [df["total_hits"].idxmax()находит индекс строки, где значение "total_hits" максимальное. Полученный индекс передается в df.loc, 
который извлечёт соответствующую строку.

Найдя нужную строку в датасете выводим основную информацию о матче: дату, домашнюю и гостевую команды, арену, общее кол-во хитов.

# 13. Выведите количество игр которая сыграла каждая команда в данном сезоне?

In [92]:
df_regular_season = df[df["season"] == "regular season"].copy()

Для анализа статистики в рамках сезона датасет предварительно фильтруем по значение "season" = "regualr season". В заданиях 13-17 будем
использовать уже отфильтрованный датасет.

In [49]:
teams = pd.concat([df_regular_season["home_team"], df_regular_season["away_team"]], ignore_index=True)
games_played = teams.value_counts().to_frame("games_played")
games_played

games_played
Kansas City Royals                      162
Philadelphia Phillies                   162
Los Angeles Dodgers                     162
Boston Red Sox                          162
Chicago White Sox                       162
San Francisco Giants                    162
St. Louis Cardinals                     162
Toronto Blue Jays                       162
Washington Nationals                    162
Colorado Rockies                        162
Chicago Cubs                            162
Houston Astros                          162
Minnesota Twins                         162
Cincinnati Reds                         162
New York Mets                           162
Seattle Mariners                        162
Milwaukee Brewers                       162
New York Yankees                        162
Oakland Athletics                       162
Pittsburgh Pirates                      162
San Diego Padres                        162
Tampa Bay Rays                          162
Texas Rangers                           162
Arizona Diamondbacks                    162
Baltimore Orioles                       162
Los Angeles Angels of Anaheim           162
Miami Marlins                           161
Detroit Tigers                          161
Atlanta Braves                          161
Cleveland Indians                       161

Поскольку каждая команда играет как дома, так и в гостях, каждая команда учитывается как в столбце "home_team", так и в столбце "away_team". \
С помошью метода pd.concat объединяем два столбца в одну последовательностьи подсчитываем кол-во появлений каждой команды.


# 14. Какая команда выиграла наибольшое количество матчей в сезоне? (Будьте внимательны с типом матча в сезоне - season).

In [93]:
home_wins = df_regular_season.groupby("home_team")["home_team_win"].sum()
away_wins = df_regular_season.groupby("away_team")["home_team_loss"].sum()
total_wins = home_wins.add(away_wins, fill_value = 0).sort_values(ascending=False) 
total_wins.head(1)

home_team
Chicago Cubs    103
dtype: int64

Победы домашней команды мы можем определить по столбцу "home_team_win", а победу гостевой по столбцу "home_team_loss", так как поражение 
домашней команды означает победу готсевой команды. В total_wins передаем сумму побед в домашних и гостивых матчах, сортируя по убывания и оставляя 
только самую верхнюю строку, которая будет выводить название команды с наибольшим кол-вом побед с сезоне.

# 15. Какая команда выиграла наибольшее количество домашних матчей в сезоне?

In [94]:
df_regular_season.groupby("home_team")["home_team_win"].sum().sort_values(ascending=False).head(1)

home_team
Chicago Cubs    57
Name: home_team_win, dtype: int64

С помощью столбца "home_team_win" подсчитываем кол-во побед каждой команды в домашних матчах регулярного сезона, сортируем по убыванию и 
выводим первую строку.

# 16. Какая команда выиграла наибольшее количество гостевых матчей в сезоне?

In [95]:
df_regular_season.groupby("away_team")["home_team_loss"].sum().sort_values(ascending=False).head(1)

away_team
St. Louis Cardinals    48
Name: home_team_loss, dtype: int64

Поражение домашней команды означает победу гоостевой, поэтому этот признак будет использоваться для учета гостевых побед.
Данный группируются по названию гостевой команды, после чего суммируется поражение домашних команд. Команды с наибольшим значением
данного показателя будет определяться как лидер по числу побед в гостевых матчах сезона.

# 17. Какая команда проиграла наибольшее количество матчей в сезоне?

In [96]:
home_losses = df_regular_season.groupby("home_team")["home_team_loss"].sum()
away_losses = df_regular_season.groupby("away_team")["home_team_win"].sum()
total_losses = home_losses.add(away_losses, fill_value=0).sort_values(ascending=False)
total_losses.head(1)

home_team
Minnesota Twins    103
dtype: int64

Для определения команды с наибольшим кол-вом поражений учитываются как домашние, так и гостевые поражения. Домашние поражения определяются
по столбцу "home_team_loss", а гостевые поражения по столбцу "home_team_win". После поражения суммируются для каждой команды и выводится команды с наибольшим числом общих поражений. 

# 18. Зависит ли выигрыш от количества посетителей матча? (https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.cov.html)

# 19. Правда ли что большинство проигрышных домашних матчей приходятся на Субботу и Воскресенье?

In [97]:
home_losses = df[df["home_team_loss"] == 1]
weekend_losses_share = home_losses["day_of_week"].isin(["Saturday", "Sunday"]).mean()
weekend_losses_share

np.float64(0.31519861830742657)

Для проверки данного утверждения рассматриваются только те матчи, в которых домашняя команда проиграла. Такие матчи отбираются с использованием признака "home_team_loss". Далее анализируется столбец "day_of_week", чтобы определить какая доля проигрышных матчей пришлась на выходные дни(субботу и воскресенье). Логическое выражение возвращает значения True/False. Среднее значение позволяет получить напрямую долю матчей, пришедшихся на выходные.

полученная доля составляет 0.315, что эквивалентно 31.5%, что <50%, а значит утверждение считается невенрным.

# 20. Правда ли что наибольшее количество ранов происходит в холодную погоду? (Холодной погодой считается погода ниже 0 градусов)

In [98]:
df["is_cold"] = df["temperature"] < 32
df.groupby("is_cold")["total_runs"].mean()

is_cold
False     8.932169
True     12.000000
Name: total_runs, dtype: float64

Вводится дополнительныйлогический признак "is_cold", который принимает значение "True", если температура воздуха ниже 0C, что эквивалентно 32F. 
После сравниваем среднее кол-во ранов ("total_runs") в матчах, проходивших в холодную и не холодную погоду. Сравнение средних значений позволяет оценить связана ил более высокая результативность с холодными погодными условиями.